src/main/java/org/springframework/integration/aws/outbound/S3MessageHandler.java

```diff
@@ -312,40 +309,64 @@ private Transfer upload(Message<?> requestMessage) {
 			if (this.uploadMetadataProvider != null) {
 				this.uploadMetadataProvider.populateMetadata(metadata, requestMessage);
 			}
-			InputStream inputStream;
-			if (payload instanceof InputStream) {
-				inputStream = (InputStream) payload;
-			}
-			else if (payload instanceof File) {
-				File fileToUpload = (File) payload;
-				if (key == null) {
-					key = fileToUpload.getName();
-				}
-				try {
-					inputStream = new FileInputStream(fileToUpload);
 
+			PutObjectRequest putObjectRequest = null;
+
+			try {
+				if (payload instanceof InputStream) {
+					InputStream inputStream = (InputStream) payload;
+					Assert.state(inputStream.markSupported(),
```

Comment:
Would be better to move this `Assert` exactly into `if (metadata.getContentMD5() == null) {`.
Otherwise it may break my logic when I obtain and populate `contentMD5` before this process and I don't care already that my `InputStream` isn't resettable.